File: 1_SOLID_Design_Patterns/1.02_Open_Close_Principle/Python_Snippet/open_close_principle.py
'''
Create a payment system to accept different types of payment
'''

# Violates Open Close Principle
class Payment(object):
    def cash_payment(self):
        print("Cash Payment Complete")
    
    def debit_card_payment(self):
        print("Debit Card Payment Complete")
    
    def credit_card_payment(self):
        print("Credit Card Payment Complete")

# in above example to add new payment methods we need to modify the class 'Payment'
        
# Complies with Open Close Principle
class Payment(object):
    def complete_payment(self):
        print("Payment Class initialised")

class CashPayment(Payment):
    def complete_payment(self):
        print("Cash Payment Complete")
    
class DebitCardPayment(Payment):
    def complete_payment(self):
        print("Debit Card Payment Complete") 

class CreditCardPayment(Payment):
    def complete_payment(self):
        print("Credit Card Payment Complete")

File: 1_SOLID_Design_Patterns/1.02_Open_Close_Principle/Java_Snippet/open_close_principle.java
// Create a payment system to accept different types of payment

//  Violates Open Close Principle

class Payment {
    public void cash_payment() {
        System.out.println("Cash Payment Complete");
    }

    public void debit_card_payment() {
        System.out.println("Debit Card Payment Complete");
    }

    public void credit_card_payment() {
        System.out.println("Credit Card Payment Complete");
    }

}

// in above example to add new payment methods we need to modify the class
// 'Payment'

// Complies with Open Close Principle

/**
 * open_close_principle
 */
interface MakePayment {
    public void complete_payment();
}

class CashPayment implements MakePayment {

    @Override
    public void complete_payment() {
        System.out.println("Cash Payment Complete");
    }

}

class DebitCardPayment implements MakePayment {

    @Override
    public void complete_payment() {
        System.out.println("Debit Card Payment Complete");
    }

}

class CreditCardPayment implements MakePayment {

    @Override
    public void complete_payment() {
        System.out.println("Credit Card Payment Complete");
    }

}
